art_tree.go

```diff
@@ -70,30 +70,29 @@ func (t *ArtTree) searchHelper(current *ArtNode, key []byte, depth int) *ArtNode
 	// While we have nodes to search
 	if current != nil {
 		maxKeyIndex := len(key) - 1
+		if depth > maxKeyIndex {
+			return current
+		}
 
-		// Check if the current is a match
-		if current.IsLeaf() {
-			if current.IsMatch(key) {
-				return current
-			}
-
-			// Bail if no match
-			return nil
-		} else if depth > maxKeyIndex {
+		// Check if the current is a match (including prefix match)
+		if current.IsLeaf() && len(current.key) >= len(key) && bytes.Equal(key, current.key[0:len(key)]) {
```

Comment:
The issue this fixes is that if you made it to a leaf node in `searchHelper` that was a prefix match, but not an exact match, the previous code wasn't returning it. (This happens trivially if you just put one short string in the index and then prefix search for it.)

The `prefixMismatch` function is confusing, which is why I changed the docstring, but I'm not sure it's wrong -- total mismatch should return `0` but matching the first character should return `1`. The problem was that for prefix searching we care about both "matched the whole prefix of the node" and "matched the whole key but had some prefix left over".

Anyway, this approach seems to be working for now, but feel free to refactor or work out another approach if you want; I'll admit this version feels a little hacky.